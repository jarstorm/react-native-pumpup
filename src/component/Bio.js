import React, { Component } from 'react'
import {
  StyleSheet,
  Text,
  View,
  Linking
} from 'react-native'
import { readMore, readLess } from '../action'
import { connect } from 'react-redux'
import { bool, object, func} from 'prop-types'

class Bio extends Component {

  renderWord(word) {
    if ('@' === word[0]) {
      return (<Text key={word} style={styles.link} onPress={() => 
        Linking.openURL('https://twitter.com/' + word.slice(1))}>{word} </Text>)
    } else if ('#' === word[0]) {
      return (<Text key={word} style={styles.link} onPress={() => 
        Linking.openURL('https://twitter.com/hashtag/' + word.slice(1))}>{word} </Text>)
    } 
    return (<Text key={word}>{word} </Text>)  
  }

  renderLine(line) {
    const parts = line.split(' ')
    return (
      <Text key={line}>
        {parts.map(word => this.renderWord(word))}
        <Text>{'\n'}{'\n'}</Text>
      </Text>
    )
  }

  parseText(text) {
    const lines = text.split('\n\n')
    return lines.map(line => this.renderLine(line))
  }

  render() {
    if (!this.props.expanded) {
      return (
        <View>
          <Text numberOfLines={3}
            ellipsizeMode='tail'
            style={{flexWrap: 'wrap'}}>
            {this.parseText(this.props.profile.bio)} 
          </Text>
          <View style={styles.actionButton}>
            <Text style={styles.actionText} onPress={this.props.readMore}>
              Read more
            </Text>
          </View>
        </View>
      )
    } 
    return (
      <View>
        <Text>{this.parseText(this.props.profile.bio)}</Text>
        <View style={styles.actionButton}>
          <Text style={styles.actionText} onPress={this.props.readLess}>
            Read less
          </Text>   
        </View>
      </View>
    )    
  }
}

Bio.propTypes = {
  expanded: bool.isRequired,
  profile: object.isRequired,
  readMore: func.isRequired,
  readLess: func.isRequired
}

const styles = StyleSheet.create({
  actionButton: {
    flex: 1, 
    flexDirection: 'row', 
    justifyContent: 'flex-end', 
    marginRight: 10
  },
  actionText: {
    color: 'blue'
  },
  link: {
    color: 'blue'
  }
})

const mapStateToProps = state => {  
  const {profile, read_more, expanded}  = state.main
  return {profile, read_more, expanded}
}

export default connect(mapStateToProps, { readMore, readLess })(Bio)

