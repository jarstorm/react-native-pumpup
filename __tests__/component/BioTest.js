// __tests__/Intro-test.js
import React from 'react'

import Bio from '../../src/component/Bio'

// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer'

import configureMockStore from 'redux-mock-store'
import expect from 'expect'
import thunk from 'redux-thunk'
import { Provider } from 'react-redux'

const mockStore = configureMockStore([thunk])


test('renders correctly', () => {
  const store = mockStore({main: {profile: {bio: ''}, expanded: false}})
  const tree = renderer.create(
    <Provider store={store}>
      <Bio />
    </Provider>
  ).toJSON()  

  expect(tree).toMatchSnapshot()
})