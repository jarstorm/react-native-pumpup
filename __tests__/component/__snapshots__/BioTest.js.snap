// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View>
  <Text
    accessible={true}
    allowFontScaling={true}
    disabled={false}
    ellipsizeMode="tail"
    numberOfLines={3}
    style={
      Object {
        "flexWrap": "wrap",
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      disabled={false}
      ellipsizeMode="tail"
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        disabled={false}
        ellipsizeMode="tail"
      >
        
         
      </Text>
      <Text
        accessible={true}
        allowFontScaling={true}
        disabled={false}
        ellipsizeMode="tail"
      >
        

        

      </Text>
    </Text>
  </Text>
  <View
    style={
      Object {
        "flex": 1,
        "flexDirection": "row",
        "justifyContent": "flex-end",
        "marginRight": 10,
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      disabled={false}
      ellipsizeMode="tail"
      onPress={[Function]}
      style={
        Object {
          "color": "blue",
        }
      }
    >
      Read more
    </Text>
  </View>
</View>
`;
